java-checks/src/main/java/org/sonar/java/checks/OneExpectedCheckExceptionCheck.java

```diff
@@ -0,0 +1,180 @@
+    .ofTypes(JUNIT4_ASSERT, "org.junit.jupiter.api.Assertions")
+    .names("fail")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.TRY_STATEMENT, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      Arguments arguments = mit.arguments();
+      if (JUNIT4_ASSERT_THROWS_WITH_MESSAGE.matches(mit)) {
+        processAssertThrowsArguments(arguments.get(1), arguments.get(2));
+      } else if (arguments.size() >= 2 && ALL_ASSERT_THROWS_MATCHER.matches(mit)) {
+        processAssertThrowsArguments(arguments.get(0), arguments.get(1));
+      }
+    } else {
+      TryStatementTree tryStatementTree = (TryStatementTree) tree;
+      if (isTryCatchIdiom(tryStatementTree)) {
+        tryStatementTree.catches().forEach(c ->
+          reportMultipleCallThrowingExceptionInTree(c.parameter().type().symbolType(), tryStatementTree.block(), c.parameter().type())
+        );
+      }
+    }
+  }
+
+  private void processAssertThrowsArguments(ExpressionTree expectedType, ExpressionTree executable) {
+    Optional<IdentifierTree> expectedException = getExpectedException(expectedType);
+    if (expectedException.isPresent() && executable.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      IdentifierTree expectedIdentifier = expectedException.get();
+      reportMultipleCallThrowingExceptionInTree(expectedIdentifier.symbolType(), ((LambdaExpressionTree) executable).body(), expectedIdentifier);
+    }
+  }
+
+  private static Optional<IdentifierTree> getExpectedException(ExpressionTree expectedType) {
+    if (expectedType.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelect = ((MemberSelectExpressionTree) expectedType);
+      ExpressionTree expression = memberSelect.expression();
+      if ("class".equals(memberSelect.identifier().name()) && expression.is(Tree.Kind.IDENTIFIER)) {
+        return Optional.of((IdentifierTree) expression);
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isTryCatchIdiom(TryStatementTree tree) {
+    List<StatementTree> statementTrees = tree.block().body();
+    if (!statementTrees.isEmpty()) {
+      StatementTree lastElement = statementTrees.get(statementTrees.size() - 1);
+      if (lastElement.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+        ExpressionTree expressionTree = ((ExpressionStatementTree) lastElement).expression();
+        if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+          return JUNIT_FAIL_MATCHER.matches((MethodInvocationTree) expressionTree);
+        }
+      }
+    }
+    return false;
+  }
+
+  private void reportMultipleCallThrowingExceptionInTree(Type expectedException, Tree tree, Tree reportLocation) {
```

Comment:
I don't see the point of the first parameter. It seems to me that `reportLocation` is a `TypeTree`, so you could just rely on it and extract the type within this method!
```
private void reportMultipleCallThrowingExceptionInTree(Tree tree, TypeTree exception) {
```
I would also rename "tree" by "treeToVisit" to make it clear